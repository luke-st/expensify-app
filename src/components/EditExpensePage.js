import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses'
import RemoveModal from './RemoveModal';

export class EditExpensePage extends React.Component {
    state = {
        promptRemove: undefined
    }
    onSubmit = (expense) => {
        this.props.startEditExpense(expense, this.props.expense.id)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.setState(() => ({ promptRemove: true }))
    }
    onRemoveConfirm = () => {
        this.props.startRemoveExpense(this.props.expense.id)
        this.props.history.push('/')
    }
    handleClearPromptRemove = () => {
        this.setState(() => ({ promptRemove: undefined }))
    }
    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='content-container'>
                        <h1 className='page-header__title'>Edit Expense</h1>
                    </div>    
                </div>
                <div className='content-container'>
                    <ExpenseForm 
                        expense={this.props.expense}
                        onSubmit={this.onSubmit} 
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
                </div>
                <div>
                <RemoveModal 
                    promptRemove={this.state.promptRemove}
                    handleClearPromptRemove={this.handleClearPromptRemove}
                    onRemoveConfirm={this.onRemoveConfirm}
                />
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (expense, id) => dispatch(startEditExpense(id, { ...expense })),
    startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)