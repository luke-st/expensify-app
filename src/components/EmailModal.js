import React from 'react'
import Modal from 'react-modal'
import { startLoginEmail, startEmailSignup } from '../actions/auth'

class EmailModal extends React.Component {
    state = {
        mode: 'LOGIN',
        error: ''
    }
    toConsole = (e) => {
        e.preventDefault()
        if (this.state.mode === 'LOGIN') {
            const email = e.target.email.value
            const password = e.target.password.value
            // console.log(email, password)
            startLoginEmail(email, password)
            .catch((e) => {
                if (e.code === 'auth/user-not-found') {
                    this.setState({ error: 'User not found. Create account or try again.'})
                } else if (e.code === 'auth/wrong-password') {
                    this.setState({error: 'You entered the wrong password. Try again.'})
                } else if (e.code === 'auth/too-many-requests') {
                    this.setState({error: 'You have attempted to login too many times. Try again later.'})
                } else if (e.code === 'auth/invalid-email') {
                    this.setState({error: 'Enter a valid email address.'})
                }
                else {
                    this.setState({error: e.code})
                }
            })
        } else if (this.state.mode === 'CREATE') {
            const email = e.target.email.value
            const password = e.target.password.value
            startEmailSignup(email, password)
            .catch((e) => {
                this.setState({error: e.code})
            })
        }
    }
    onModeChange = (e) => {
        const mode = e.target.value
        this.setState({ mode })
    }
    render() {
        return (
            <Modal
                isOpen={!!this.props.emailPrompt}
                onRequestClose={this.props.handleClearEmailPrompt}
                contentLabel="Email Sign up"
                closeTimeoutMS={200}
                className='modal'
                appElement={app}
            >
                <h3 className='modal__title'>Please enter your email and chosen password.</h3>
                {this.state.error && <p className='form__error'>{this.state.error}</p>}
                <div>
                    <div className="input-group modal-input">
                    <select className="button"
                        onChange={this.onModeChange}
                    >
                        <option value="LOGIN">Login</option>
                        <option value="CREATE">Create Account</option>
                    </select>
                    </div>
                    <form className="form modal-input"
                        // onSubmit={props.startLoginEmail(this.state.mode, {})}
                        onSubmit={this.toConsole}
                    >
                        <input type='email' id='email' placeholder='you@example.com' autoComplete='true' />
                        <input type='password' id='password' placeholder='Password' minLength='6' autoComplete='true' />
                        <button>Submit</button>
                    </form>
                </div>
            </Modal>
        )
    }
}

export default EmailModal