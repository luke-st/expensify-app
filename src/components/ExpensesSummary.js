import React from 'react'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'
import { connect } from 'react-redux'
import numeral from 'numeral'

export const ExpensesSummary = (props) => (
    <div>
    <h3>Viewing {props.expenseCount} {props.expenseCount === 1 ? 
            'expense' 
            : 
            'expenses'} totalling {props.expensesTotal}</h3>
    </div>
)

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: numeral(selectExpensesTotal(visibleExpenses) / 100).format('$0,0.00')
    }
}

export default connect(mapStateToProps)(ExpensesSummary)