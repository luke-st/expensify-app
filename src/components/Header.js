import React from 'react';
import { NavLink } from "react-router-dom";
const Header = () => (
    <header>
        <h1>Expensify</h1>
        <div className="nav-bar">
            <NavLink to="/" exact activeClassName="is-active">Dashboard</NavLink>
            |
            <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        </div>
    </header>
);

export default Header