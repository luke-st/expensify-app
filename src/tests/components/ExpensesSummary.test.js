import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'

test('Should render ExpensesSummary correctly with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} expensesTotal={'£1,141.95'} expenseCount={3} />)
    expect(wrapper).toMatchSnapshot()
})

test('Should render ExpensesSummary correctly with only 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[2]]} expensesTotal={'£45.00'} expenseCount={1} />)
    expect(wrapper).toMatchSnapshot()
})

test('Should render when no expenses are provided', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[]} expensesTotal={'£0.00'} expenseCount={0} />)
    expect(wrapper).toMatchSnapshot()
})