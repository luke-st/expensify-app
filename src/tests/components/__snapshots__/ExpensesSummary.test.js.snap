// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary correctly with multiple expenses 1`] = `
<div>
  <h3>
    Viewing 
    3
     
    expenses
     totalling 
    £1,141.95
  </h3>
</div>
`;

exports[`Should render ExpensesSummary correctly with only 1 expense 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
     totalling 
    £45.00
  </h3>
</div>
`;

exports[`Should render when no expenses are provided 1`] = `
<div>
  <h3>
    Viewing 
    0
     
    expenses
     totalling 
    £0.00
  </h3>
</div>
`;
